### PR TITLE
Add `finished` signal to GPUParticles

### LINKED ISSUE #54964: Vulkan GPUParticles3D related error spam with multithreaded renderer
### Godot version

4.0.dev (19571c9c4b)

### System information

W10, Vulkan, RTX2060

### Issue description

When using GPUParticles3D in a scene with `driver/threads/thread_model=2` (Multithreaded render mode), this error is printed repeatedly:
```cpp
E 0:00:07:0555   RendererStorageRD::particles_is_inactive: This function should never be used with threaded rendering, as it stalls the renderer.
  <C++ Error>    Condition "RenderingServerGlobals::threaded" is true. Returning: false
  <C++ Source>   servers\rendering\renderer_rd\renderer_storage_rd.cpp:5697 @ RendererStorageRD::particles_is_inactive()
```

The only place this could be called every frame is here:
https://github.com/godotengine/godot/blob/16dc607f8ad2055c8dec543bcd427fe74e5d6059/servers/rendering/renderer_scene_cull.cpp#L2700

### Steps to reproduce

Create a 3D scene with GPUParticles3D set up.
Enable multithreaded rendering mode.
Run project, see errors.

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by HolonProduction**
Followup to #76853.

_bugsquad edit: Fixes: https://github.com/godotengine/godot/issues/54964_

### CODE REVIEW

**Comment by mr-sihc** [comment]
when?
threaded render ist still broken without this

**Comment by Calinou** [comment]
> when? threaded render ist still broken without this

4.1 is in feature freeze, and this pull request needs a review before this can be merged.

File: scene/2d/gpu_particles_2d.cpp
Comment (clayjohn) [inline scene/2d/gpu_particles_2d.cpp:48]: This introduces a significant change in behaviour. Is it necessary for this feature to work? 

Notably, it makes the behaviour of ``set_emitting()`` differ between CPUParticles and GPUParticles. So if this change is necessary, it needs to be made in both forms of particles

Comment (HolonProduction) [reply scene/2d/gpu_particles_2d.cpp:48]: The usage of `set_emitting` for one shot particle systems feels very unintended in my opinion. The resetting of the property and the time it takes till it can be done again feels wrong. It does not respect explosivness and I thought restarting would make sense. Moreover it saved some code copying for the timer reset. But you are right that this is a change of behaviour that shouldn't just be introduced.

With that said it seems nearly impossible to support the signal when using `set_emitting` without restarting. Setting the `emitting` property throws of the time prediction and in contrast to CPUParticles we can't just access the particles. So the other option I see is to only support the signal when `restart` is used. When using `one_shot` this method is the best option anyway. As long as it is documented, that usage of `emitting` can throw of the signal this seems ok to me.

For consistency the CPU particles could be changed as well to only support the signal with `restart`. But since the usage of emitting works fine there this discrepancy would be fine IMHO.

Comment (clayjohn) [reply scene/2d/gpu_particles_2d.cpp:48]: I think the ideal solution would be to maintain the same behaviour as with CPUParticles in this PR (i.e. not reset in `set_emitting()` and then make another PR that adds the reset to `set_emitting()` for both GPUParticles and CPUParticles. I know the behaviour of setting the ``emitting`` property with `one_shot` is a common pain point for users, but I don't think this is the right place to address that pain point. especially since it breaks compatibility and creates a difference in behaviour between CPUParticles and GPUParticles. 

From my understanding, the problem with not restarting in ``set_emitting()`` for the signal is that the signal will fire when the particles _should_ have finished anyway right? even though they technically never finished. If that is the case, then couldn't we check if ``!p_emitting && one_shot && emitting`` and then in that case set a flag to cancel the signal?

Comment (HolonProduction) [reply scene/2d/gpu_particles_2d.cpp:48]: Actually since internal process is turned of when `!p_emitting` the problems would only occur if you were to turn off emitting and then turn it on again but at this point I have the feeling im overthinking the thing a little bit.

We should just reset the timer everytime emitting is turned on but not reset the gpu particles. The particle behaviour stays the same and the signal will always fire after the particles finished (though with some delay).

Comment (HolonProduction) [reply scene/2d/gpu_particles_2d.cpp:48]: That is correct. But at least it does not break existing behaviour. I just see no way to make an accurate estimate of remaining time when emitting is set to true without syncing the server. So either we say it will be late or we scrap the signal completly if restart is not used.

Comment (clayjohn) [reply scene/2d/gpu_particles_2d.cpp:48]: Well, right now if emitting is set to false and then true before the full cycle has ended, it essentially just gets ignored. So I would do the same on the node side and just ignore it. 

IMO with the signal, we will already be a huge step up over the behaviour in master. Previously users would often set ``emitting`` to true while the cycle was going in the hopes that the cycle would continue or restart. Since that didn't work, they had to guess when the particle system finished so they could set emitting to true to start it again. With this PR, we can tell them just to wait for the signal before setting ``emitting`` to true. Alternatively, they can call ``restart()`` to restart immediately.

**Review by clayjohn — DISMISSED** [review]
This current iteration looks good to me. I have tested locally with GPUParticles2D and GPUParticles3D and the behaviour appears to match CPUParticles (with https://github.com/godotengine/godot/pull/76853)

As we discussed, it would be worth opening a new proposal/PR to deal with the awkwardness of particles not resetting when emitting is toggled off then on again.

Note for mergers, this PR should be merged at the same time as https://github.com/godotengine/godot/pull/76853

**Comment by clayjohn** [comment]
One thing I just realized after hitting "approve". Is "emission_finished" the best name for this signal? As it fires after the last particle dies rather than after the last particle is emitted. The way I see it, there are two stages for a one_shot particle 1) Particles emitting (and some active) and 2) no particles emit anymore, but some particles still active. This is fired after 2), but "emission_finished" seems to infer that it signals between 1) and 2)

It would be nice to have feedback from the particles team CC @QbieShay

**Comment by QbieShay** [comment]
I'd maybe just call it "finished" or "cycle done". What happens if it's not oneshot? will it still fire at the end of each cycle of emission?

**Comment by HolonProduction** [comment]
Im open to renaming the signal. I just took the name from the old implementation in #7887. The signal is one shot only so `finished` seems fitting.

**Comment by clayjohn** [comment]
I like "finished". "cycle_done" seems like it would fire every time a cycle finished (which would match this behaviour for one_shot particles, but would fire many times for non one_shot particles).

Let's do two things then:
1. Rename the signal to "finished" ("finished"already exists as a SceneStringName, so you can just use it and delete "emission_finished"
2. Add to the docs to clarify when the signal is fired (e.g. "emitted when all active particles have finished processing. When [member one_shot] is disabled, particles will process continuously, so this is never emitted")

**Comment by QbieShay** [comment]
Do you think it's possible to have the signal emit also for non-oneshot particles? I can see usage of this for baking flipbook animations for lower LOD vfx

**Comment by QbieShay** [comment]
The actual implementation doesn't need to be in this PR but the name would need to include this usecase.

**Comment by HolonProduction** [comment]
> Do you think it's possible to have the signal emit also for non-oneshot particles? I can see usage of this for baking flipbook animations for lower LOD vfx

Should be doable but would propably require a different approach for CPU Particles (GPU should work very similar). But this could be put behind a condition.

It is a thing for another PR though. And I would prefer to resolve the emmitting restart for one shot before it.

Im not very sure of what to think with `cycle_done` personally. It feels kinda wrong at this point with only `one_shot` supported. But if we assume, that continuous support is going to be added eventually it might be justified.

**Comment by clayjohn** [comment]
> Do you think it's possible to have the signal emit also for non-oneshot particles? I can see usage of this for baking flipbook animations for lower LOD vfx

Are you thinking a of a signal that emits every cycle?

**Comment by QbieShay** [comment]
I'm not so sure about the names either. I'd like something that encompasses both usecases, which are:

The particle system has a set number X of particles, that are emitted at different intervals depending on the explosiveness and randomness parameters. There's multiple concepts in this
1. I have emitted all particles
2. All particles i have emitted in this cycle are now done processing and have disappeared. New ones may be there already
3. I have finished "my job" because once i do one emission cycle, I do not need to do it again.

I need some more time to come up with a name that encompasses 3 and 2, but I think `emission_finished` hints more at usecase 1 rather than 2 and 3, while `finished` is more generic for "finished for whatever i set out to do" and i think fits more with 2 and 3. `cycle_finished` is more precise but also "cycle" would be used only here. I do have plans in the future to add properties over emitter cycle (velocity multiplier for example, imagine water that shoots out at high pressure in the beginning and it's reduced to a drip in the end), but that would require a specific definition of emission cycle. It would also be interesting to add a pause between one cycle of emission and the next.

I'm sharing all of this to give a broader vision on how I'd like particles to be developed in the future. I hope this is useful to pick a name. I would just not use `emission_finished` because it hints at case 1 while we're tackling 2 and 3 with this signal.

**Comment by HolonProduction** [comment]
I guess I will go with `cycle_finished` then.

**Comment by clayjohn** [comment]
Thanks for the explanation. Using the numbers from your response. 

1. would be a good candidate for "emission_finished" (but I'm not sure if it would be useful)
2. could be a good candidate for "cycle_finished" but indeed. It doesn't really match exactly ``cycle`` is an internal variable we use to track how many times an emission "cycle" has completed. I.e., If I emit all my particles 1..10, once 1. dies and I emit it again, I am on cycle 2 and so on. Each time particle 1 get's re-emitted, we increment the cycle. So a cycle in this terminology doesn't quite fit what you describe in use-case 2.
3. is exactly what this PR covers, and I think is a good candidate for the name "finished"

**Comment by QbieShay** [comment]
@HolonProduction let us reach a consensus first before updating the PR. I don't want to make you run around and changed it multiple times, I don't think it would be a good use of your time.

I know it looks like annoying bikeshedding but i think it's important to consider future expansion to keep names somewhat consistent and communicative.

**Comment by HolonProduction** [comment]
> @HolonProduction let us reach a consensus first before updating the PR.

Totally agree. My work on it is on hold until there is a agreement.

**Comment by QbieShay** [comment]
So, I do personally think that the most common usecase will be to have a oneshot particle that you want to auto-free once it's done. I think that can have a more default name like `finished`.

For consistency, what are the other nodes that have a concept of being "done"? Animators have `animation_finished`, and what else?

**Comment by lawnjelly** [comment]
"particles_finished"? :thinking:

**Comment by clayjohn** [comment]
> For consistency, what are the other nodes that have a concept of being "done"? Animators have animation_finished, and what else?


Here are the currently defined finished signals (note emission finished is not used anywhere):
https://github.com/godotengine/godot/blob/b8ed596769d2114797015833ce86f86ee872ecfa/scene/scene_string_names.cpp#L59-L61


``finished`` appears to only be used in one place, in the video stream player

**Comment by QbieShay** [comment]
Clarification: my proposal that everything has a "finished" signal is out of the scope of this PR.

**Comment by HolonProduction** [comment]
But we could start by naming this one `finished`.

**Comment by clayjohn** [comment]
My ideal resolution would be:
1. rename "emission_finished" to just "finished" (delete "emission_finished" from SceneStringName)
2. Open proposals for other types of signals (e.g. signal on cycle, signal on emission finishing, etc.)

My only concern is the name "finished" might imply that it will signal whenever the particle system finishes, so some users may expect it to fire after a user turns emitting to false and the last particle dies. But I may be overthinking things at this point

**Comment by clayjohn** [comment]
> If that's the case we can find a way to emit the signal at that point too. I think it would make a lot of sense, but it doesn't need to be this PR.

Fair enough. We should document that this only gets emitted with ``one_shot`` particles, and then we can remove that limitation if there is demand in the future :)

@HolonProduction To summarize, we have circled back to my earlier comment https://github.com/godotengine/godot/pull/76859#issuecomment-1630731991 if you implement the two things there in this PR and the CPUParticles PR, we should be ready to merge :)

**Review by clayjohn — APPROVED** [review]
Both this and #76853 should be ready to merge now. 

One note, since they both delete "emission_finished" from SceneStringNames, github require rebasing one of the PRs after the other has been merged. But it also might be able to gracefully handle this type of conflict

**Comment by YuriSizov** [comment]
Thanks!

> But it also might be able to gracefully handle this type of conflict

To my surprise it did!

**Comment by Gatada** [comment]
In 4.2 beta 1 should the ```finished``` signal from a one shot emitter allow me to immediately set ```emitting = true``` to get new particles? Cause that is not working in my test project without manually delaying for a bit longer.

Maybe this is more related to #79689?

**Comment by clayjohn** [comment]
@Gatada In short, no. A signal is just a notification that something has happened. It doesn't impact the ``emitting`` property in any way. 

You seem to be describing desired functionality that is the subject of this proposal https://github.com/godotengine/godot-proposals/issues/7322, I suggest you follow up there

**Comment by Gatada** [comment]
I suspect I wasn't clear: the finished signal is transmitted at what time? And has it been implemented in beta 1?

**Comment by Gatada** [comment]
> @Gatada In short, no. A signal is just a notification that something has happened. It doesn't impact the ``emitting`` property in any way. 
> 
> 
> 
> You seem to be describing desired functionality that is the subject of this proposal https://github.com/godotengine/godot-proposals/issues/7322, I suggest you follow up there
> 
> 

Surely the signal isn't emitted without the emitter having changed state? The finished signal isn't very useful if I still risk interrupting or interfering with the particles life-span even after having received it.

The docs states: "Emitted when all active particles have finished processing."

Which I assume means:

• emitting is false
• all particles have expired
• emitter is ready to emit

This however is not the case for beta 1, hence my question: is my assumption wrong?

**Comment by HolonProduction** [comment]
**tl;dr**
Try using `restart()` instead of `emitting=true`.

Since we are talking about GPU particles here the answer is a little bit complicated. Those particles are processed on the GPU and querying the GPU every frame to find out whether the particles are still processing is to performance heavy. The signal is just fired based on a timer with a time estimated from you particle configuration.

However I don't think you issue has something to do with this PR. While I was working on this I also had problems with restarting. If I recall correctly setting `emitting=true` would only work after a certain time. Which is basically the time your particles need but it does not factor in explosiveness or lifetime randomness. So when using those properties (which I suspect you are doing) there is a time when no particles are visible but restarting this way does not work (idk why just observed the fact; no guaranty has been some time since I worked on this).

Setting `emitting` to `true` has no effect if some internals think the system is still running, which I suspect is happening. On the other hand `restart` tells all internals to restart the system from the start, discarding all previous particles. So using `restart` should solve the problem. In theory it could also remove existing particles but since `finished` is emitted when no particles are visible anymore this is not a concern in this case.

**Comment by Gatada** [comment]
> Documentation should reflect this.

Made a doc PR to explain the signal more accurately.

**Pull request merged**
